pallets/msa/src/offchain_storage.rs

```diff
@@ -0,0 +1,464 @@
+		}
+	} else {
+		log::info!("skipping processing events on {:?} due to existing lock!", block_number);
+	};
+}
+
+/// Set offchain index value, used to store MSA Events to be process by offchain worker
+fn set_offchain_index<V>(key: &[u8], value: V)
+where
+	V: Encode + Clone + Decode + Eq + Debug,
+{
+	offchain_index::set(key, value.encode().as_slice());
+}
+
+/// Get offchain index value, used to store MSA Events to be process by offchain worker
+fn get_offchain_index<V>(key: &[u8]) -> Option<V>
+where
+	V: Encode + Clone + Decode + Eq + Debug,
+{
+	let value = offchain_common::get_index_value::<V>(key);
+	value.unwrap_or_else(|e| {
+		log::error!("Error getting offchain index value: {:?}", e);
+		None
+	})
+}
+
+/// Offchain indexed compatible Event type
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebugNoBound)]
+enum IndexedEvent<T: Config> {
+	/// A new Message Service Account was created with a new MessageSourceId
+	IndexedMsaCreated {
+		/// The MSA for the Event
+		msa_id: MessageSourceId,
+
+		/// The key added to the MSA
+		key: T::AccountId,
+	},
+	/// An AccountId has been associated with a MessageSourceId
+	IndexedPublicKeyAdded {
+		/// The MSA for the Event
+		msa_id: MessageSourceId,
+
+		/// The key added to the MSA
+		key: T::AccountId,
+	},
+	/// An AccountId had all permissions revoked from its MessageSourceId
+	IndexedPublicKeyDeleted {
+		/// The MSA for the Event
+		msa_id: MessageSourceId,
+		/// The key no longer approved for the associated MSA
+		key: T::AccountId,
+	},
+}
+
+impl<T: Config> IndexedEvent<T> {
+	/// maps a pallet event to indexed event type
+	pub fn map(event: &Event<T>, msa_id: MessageSourceId) -> Option<Self> {
+		match event {
+			Event::MsaCreated { msa_id, key } =>
+				Some(Self::IndexedMsaCreated { msa_id: *msa_id, key: key.clone() }),
+			Event::PublicKeyAdded { msa_id, key } =>
+				Some(Self::IndexedPublicKeyAdded { msa_id: *msa_id, key: key.clone() }),
+			Event::PublicKeyDeleted { key } =>
+				Some(Self::IndexedPublicKeyDeleted { msa_id, key: key.clone() }),
+			_ => None,
+		}
+	}
+}
+
+/// Initializes the last_process_block value in offchain DB
+fn init_last_processed_block<T: Config>(current_block_number: BlockNumberFor<T>) {
+	let mut last_processed_block_lock = StorageLock::<'_, Time>::with_deadline(
+		LAST_PROCESSED_BLOCK_LOCK_NAME,
+		Duration::from_millis(LAST_PROCESSED_BLOCK_LOCK_TIMEOUT_EXPIRATION),
+	);
+	let _ = last_processed_block_lock.lock();
+	let last_processed_block_storage =
+		StorageValueRef::persistent(LAST_PROCESSED_BLOCK_STORAGE_NAME);
+
+	// setting current_block-1 as the last processed so that we start indexing from current_block
+	let target_block: BlockNumberFor<T> =
+		current_block_number.saturating_sub(BlockNumberFor::<T>::one());
+	last_processed_block_storage.set(&target_block);
+}
+
+/// Add a key into MSA offchain DB
+fn add_key_to_offchain_msa<T: Config>(key: T::AccountId, msa_id: MessageSourceId) {
```

Comment:
Suggestion: I was confused by having both this and the `process_offchain_events`

- Idea 1: Could we just construct the event and run it through `process_offchain_events`?
- Idea 2: Rename / comment this to better reflect that this is dealing with it from on-chain storage.